### ISSUE TITLE
Ecosystem upgrade megathread

### ISSUE THREAD

**Issue by kubukoz** [typelevel/cats-effect#1330]
Here, we can track the progress of major libraries upgrading to CE3.

That should be a nice way of finding out whether any of them are going to be late with the upgrade, or straight-up impossible to migrate in a relatively source-compatible fashion.

| Project                                                                           | Progress                                                                                                                                  | Blockers                                                                      |
| --------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------- |
| [Weaver](https://github.com/disneystreaming/weaver-test)                          | [released](https://github.com/disneystreaming/weaver-test/releases/tag/v0.6.3)                                                            | none                                                                          |
| [fs2](https://github.com/typelevel/fs2)                                           | [released](https://github.com/typelevel/fs2/releases/tag/v3.0.4)                                                                          | none                                                                          |
| [http4s](https://github.com/http4s/http4s)                                        | [released](https://github.com/http4s/http4s/releases/tag/v0.23.0-M1)                                                                      | none                                                                          |
| [fs2-kafka](https://github.com/fd4s/fs2-kafka)                                    | [released](https://github.com/fd4s/fs2-kafka/releases/tag/v2.0.0)                                                                         | none                                                                          |
| [fs2-grpc](https://github.com/fiadliel/fs2-grpc)                                  | [released](https://github.com/typelevel/fs2-grpc/releases/tag/v2.1.2)                                                                     | none                                                                          |
| [munit-cats-effect](https://github.com/typelevel/munit-cats-effect)               | [released](https://github.com/typelevel/munit-cats-effect/releases/tag/v1.0.3)                                                            | none                                                                          |
| [fs2-rabbit](https://github.com/profunktor/fs2-rabbit)                            | [released](https://github.com/profunktor/fs2-rabbit/releases/tag/v4.0.0)                                                                  | none                                                                          |
| [natchez](https://github.com/tpolecat/natchez)                                    | [released](https://github.com/tpolecat/natchez/releases/tag/v0.1.4)                                                                       | none                                                                          |
| [cats-effect-testing](https://github.com/djspiewak/cats-effect-testing)           | [released](https://github.com/typelevel/cats-effect-testing/releases/tag/v1.1.0)                                                          | none                                                                          |
| [cats-stm](https://github.com/TimWSpence/cats-stm)                                | [released](https://github.com/TimWSpence/cats-stm/releases/tag/v0.10.3)                                                                   | none                                                                          |
| [slick-effect](https://github.com/kubukoz/slick-effect)                           | [released](https://github.com/kubukoz/slick-effect/releases/tag/v0.4.0)                                                                   | `Async[DBIO]` support dropped                                                 |
| [Eff](https://github.com/atnos-org/eff)                                           | [released](https://github.com/atnos-org/eff/releases/tag/v5.16.0)                                                                         | unknown                                                                       |
| [redis4cats](https://github.com/profunktor/redis4cats)                            | [released](https://github.com/profunktor/redis4cats/releases/tag/v1.0.0-RC3)                                                              | none                                                                          |
| [log4cats](https://github.com/typelevel/log4cats)                                 | [released](https://github.com/typelevel/log4cats/releases/tag/v2.1.1)                                                                     | none                                                                          |
| [skunk](https://github.com/tpolecat/skunk)                                        | [released](https://github.com/tpolecat/skunk/releases/tag/v0.1.2)                                                                         | none                                                                          |
| [doobie](https://github.com/tpolecat/doobie)                                      | [released](https://github.com/tpolecat/doobie/releases/tag/v1.0.0-M4)                                                                     | `Async[ConnectionIO]` support dropped                                         |
| [Ciris](https://github.com/vlovgr/ciris)                                          | [released](https://github.com/vlovgr/ciris/releases/tag/v2.0.0-RC3)                                                                       | none                                                                          |
| [Decline](https://github.com/bkirwi/decline)                                      | [released](https://github.com/bkirwi/decline/releases/tag/v2.0.0-RC1)                                                                     | none                                                                          |
| [ZIO](https://github.com/zio/interop-cats)                                        | [released](https://github.com/zio/interop-cats/releases/tag/v3.1.1.0)                                                                     | none                                                                          |
| [Sttp](https://github.com/softwaremill/sttp)                                      | [released](https://github.com/softwaremill/sttp/releases/tag/v3.3.4)                                                                      | none                                                                          |
| [Pure Config](https://github.com/pureconfig/pureconfig)                           | [released](https://github.com/pureconfig/pureconfig/releases/tag/v0.15.0)                                                                 | none                                                                          |
| [elastic4s](https://github.com/sksamuel/elastic4s)                                | [released](https://github.com/sksamuel/elastic4s/releases/tag/v7.12.1)                                                                    | none                                                                          |
| [fs2-data](https://github.com/satabin/fs2-data)                                   | [released](https://github.com/satabin/fs2-data/releases/tag/v1.0.0-RC2)                                                                   | none                                                                          |
| [trace4cats](https://github.com/janstenpickle/trace4cats)                         | [released](https://github.com/janstenpickle/trace4cats/releases/tag/v0.12.0-RC1)                                                          | none                                                                          |
| [sup](https://github.com/kubukoz/sup)                                             | [released](https://github.com/kubukoz/sup/releases/tag/v0.9.0-M2)                                                                         | none                                                                          |
| [Scala Cache](https://github.com/cb372/scalacache/)                               | [PR merged](https://github.com/cb372/scalacache/pull/510), [published for M5](https://github.com/cb372/scalacache/releases/tag/v1.0.0-M2) | none                                                                          |
| [prox](https://github.com/vigoo/prox)                                             | [PR merged](https://github.com/vigoo/prox/pull/217)                                                                                       | none                                                                          |
| [streamz](https://github.com/davenverse/epimetheus)                               | [PR merged](https://github.com/davenverse/epimetheus/pull/188)                                                                            | none                                                                          |
| [odin](https://github.com/valskalla/odin)                                         | [PR merged](https://github.com/valskalla/odin/pull/273)                                                                                   | none                                                                          |
| [catbird](https://github.com/travisbrown/catbird/pull/267)                        | [PR for 3.0](https://github.com/travisbrown/catbird/pull/267)                                                                             | unknown                                                                       |
| [streamz](https://github.com/krasserm/streamz)                                    | [PR for M3](https://github.com/krasserm/streamz/pull/80)                                                                                  | none                                                                          |
| [tapir](https://github.com/softwaremill/tapir)                                    | [PR for 3.0](https://github.com/softwaremill/tapir/pull/1154)                                                                             | catbird                                                                       |
| [scala-opentracing](https://github.com/Colisweb/scala-opentracing)                | [PR for 3.0](https://github.com/Colisweb/scala-opentracing/pull/10)                                                                       | unknown                                                                       |
| [circe-fs2](https://github.com/circe/circe-fs2)                                   | [PR open](https://github.com/circe/circe-fs2/pull/232)                                                                                    | not expected                                                                  |
| [caliban](https://github.com/ghostdogpr/caliban)                                  | [PR open](https://github.com/ghostdogpr/caliban/pull/891)                                                                                 | unknown                                                                       |
| [natchez-http4s](https://github.com/tpolecat/natchez-http4s)                      | [PR open](https://github.com/tpolecat/natchez-http4s/pull/8)                                                                              | unknown                                                                       |
| [tofu](https://github.com/tofu-tf/tofu)                                           | [issue exists](https://github.com/tofu-tf/tofu/issues/536)                                                                                | unknown                                                                       |
| [finch](https://github.com/finagle/finch)                                         | [issue exists](https://github.com/finagle/finch/issues/1319)                                                                              | unknown                                                                       |
| [davenverse libraries](https://github.com/Davenverse)                             | varying                                                                                                                                   | not expected                                                                  |
| [http4s-micrometer-metrics](https://github.com/ovotech/http4s-micrometer-metrics) | none                                                                                                                                      | unknown                                                                       |
| [Monix](https://github.com/monix/monix)                                           | work in progress                                                                                                                          | unknown                                                                       |
| [Monix BIO](https://github.com/monix/monix-bio)                                   | unknown                                                                                                                                   | unknown                                                                       |
| [meow-mtl](https://github.com/oleg-py/meow-mtl)                                   | none                                                                                                                                      | should [update mtl to 1.x first](https://github.com/oleg-py/meow-mtl/pull/18) |
| [console4cats](https://github.com/profunktor/console4cats/issues/175)             | [Deprecated](https://github.com/profunktor/console4cats/issues/175)                                                                       | none                                                                          |

If you have edit rights, feel free to add more details / libraries.

**Comment by benkio** [comment]
I came here from your tweet 😅
In cats, I did the following doc PR to collect the libs/projects related to the typelevel ecosystem

https://github.com/typelevel/cats/pull/3609

It may be relevant to this thread I guess 
Cheers 😉

**Comment by Daenyth** [comment]
> Are you using cats-effect in production? Let us know what libraries from the ecosystem you're using

In addition to the things mentioned in the above comment;

- https://github.com/typelevel/scalacheck-effect
- https://github.com/satabin/fs2-data
- https://github.com/circe/circe-fs2
- https://github.com/ovotech/http4s-micrometer-metrics
- https://github.com/ChristopherDavenport/log4cats
- https://github.com/ChristopherDavenport/fuuid / fuuid-doobie
- https://github.com/profunktor/fs2-rabbit/
- https://github.com/ChristopherDavenport/cormorant
- https://github.com/kubukoz/slick-effect
- streamz - WIP at https://github.com/krasserm/streamz/pull/80/files

**Comment by kubukoz** [comment]
@Daenyth added. @ChristopherDavenport -verse is a vital part of it!

btw. you have edit rights, feel free to add them to OP

**Comment by mpilquist** [comment]
Can we roll fuuid in to cats-effect-std module?

**Comment by kubukoz** [comment]
> Can we roll fuuid in to cats-effect-std module?

I'm all for it. Also a small `Random[F]`, Chris's `unique` and https://github.com/ChristopherDavenport/system-effect/blob/master/src/main/scala/io/chrisdavenport/system/effect/Environment.scala

**Comment by ahjohannessen** [comment]
- [fs2-grpc](https://github.com/fiadliel/fs2-grpc) has a CE/FS2 3.0 [branch](https://github.com/fiadliel/fs2-grpc/tree/ce-fs2-3.0) with all tests green :)

**Comment by kubukoz** [comment]
Added, thanks!

**Comment by RaasAhsan** [comment]
Can we open issues for each library/class we'd like to port to `std`?

**Comment by ChristopherDavenport** [comment]
Random - https://github.com/christopherdavenport/random
Cats-Effect-Time - https://github.com/ChristopherDavenport/cats-effect-time (scalajs artifact size complications)

**Comment by kubukoz** [comment]
@RaasAhsan I like the idea, I can create some later on, unless someone beats me to it

**Comment by ghostdogpr** [comment]
For ZIO interop, the repo is this one https://github.com/zio/interop-cats

**Comment by jbwheatley** [comment]
[fs2-rabbit](https://github.com/profunktor/fs2-rabbit) has a CE/FS2 3.0 [branch](https://github.com/profunktor/fs2-rabbit/tree/ce3) with all tests green

**Comment by felixbr** [comment]
`catbird` seems to be missing from the list (Finch is using it for example).

I tried to implement the new typeclasses but I'm again blocked by cancelation requirements. My thoughts are here: https://github.com/travisbrown/catbird/pull/267

**Comment by felixbr** [comment]
[redis4cats](https://github.com/profunktor/redis4cats ) - no progress so far it seems

**Comment by kubukoz** [comment]
A note for anyone passing by, if you want to help, there's no official migration guide yet, but some useful information is being gathered in https://github.com/typelevel/cats-effect/issues/1048 - should be a good place to start.

**Comment by djspiewak** [comment]
@kubukoz If you've got some time to spare, do you think you could chase down current status on some of this? It'd be good to get a table going in the OP to indicate status and blockers if there are any.

**Comment by TimWSpence** [comment]
https://github.com/TimWSpence/cats-stm is running on CE3 :)

**Comment by kubukoz** [comment]
oof, thanks for the reminder @djspiewak - I'll do the rounds now.

**Comment by kubukoz** [comment]
Ney @felixbr, are you planning on updating the catbird repository to the latest milestones (no rush though)? I can't comment there, since I'm blocked 😅

**Comment by kubukoz** [comment]
Added a table to clean this all up, now thinking of creating a project to track this instead - it'll be easier to spot which projects need attention. Thoughts?

**Comment by djspiewak** [comment]
The table is great! Also I can report that munit-cats-effect already releases for CE3, which is good because it has a bunch of downstream dependencies. In that vein, cats-effect-testing is also now also fully updated.

**Comment by kubukoz** [comment]
Oh, it releases for both. I guess my cmd+f only showed me the last version here and I gave up. Good, good. Thanks!

![image](https://user-images.githubusercontent.com/894884/107158107-2e687b00-6988-11eb-97fb-c774ca3ee720.png)

**Comment by djspiewak** [comment]
Yeah the munit-cats-effect build is unique to my knowledge. The only one remotely close to it is Weaver. Most projects are just maintaining separate series branches because so much source has to be changed.

**Comment by felixbr** [comment]
> Ney @felixbr, are you planning on updating the catbird repository to the latest milestones (no rush though)?

I will try to update the PR from M2 to M5. It will probably not be ready for a merge into master for a while until we decide how to go about making the switch. We also haven't checked yet if my PR still covers all use-cases of downstream libraries (e.g. `finch`).

**Comment by felixbr** [comment]
I'm a bit concerned that `Monix` doesn't even seem to have a WIP migration to CE3. It's one of the few actual implementers of the full CE spec and finalizing CE3 without seeing if other implementers can follow suite with `IO` would not be ideal in my mind. If a library like `Monix` could not implement CE3 fully, it would also lose access to (some of) the `fs2` ecosystem.

Not trying to push anyone, just thinking out loud. 

For `catbird` we might have to drop certain CE features because we cannot implement the full spec (neither for CE2 nor CE3) and the requirements to be compliant/lawful seem even stricter now in some cases.

**Comment by djspiewak** [comment]
> I'm a bit concerned that Monix doesn't even seem to have a WIP migration to CE3. It's one of the few actual implementers of the full CE spec and finalizing CE3 without seeing if other implementers can follow suite with IO would not be ideal in my mind. If a library like Monix could not implement CE3 fully, it would also lose access to (some of) the fs2 ecosystem.

Agreed. I think the blocker there has mostly been people's time and availability. It is believed that the major new elements will be straightforward to implement, but no one has.

> and the requirements to be compliant/lawful seem even stricter now in some cases.

They definitely are. Kind of sucks for projects like catbird, but it makes the ecosystem a bit tighter and more generic, which was the goal.

**Comment by kubukoz** [comment]
Note to self: we're missing on the list:

- softwaremill/tapir
- odin
- weaver

 
will update the list and ping the maintainers later

**Comment by bplommer** [comment]
fs2-kafka now has a milestone release (v3.0.0-M1) on RC2.

**Comment by kubukoz** [comment]
I'll probably do a round of updates in the post this weekend :) seeing some commits in these repos in my inbox every day!

**Comment by kubukoz** [comment]
Most projects seem to be doing fine (and have updated to RC2) :)

**Comment by felixbr** [comment]
Thanks to `MonadCancel.Uncancelable` it looks like it will be possible for `catbird` to support the same feature set as with `cats-effect` 2.x. 🙂

I've updated the PR to use `3.0.0-RC2` and the laws pass.

**Comment by kubukoz** [comment]
the table is growing! Added some tpolecat and tracing libs.

**Comment by kubukoz** [comment]
Note to self, add:

```
trace4cats status:
progress: PR for RC2: janstenpickle/trace4cats#324
blockers: zio/interop-cats, sttp, tapir
```

thanks for the update @catastrophe

---

Also, we're at RC3 now - I might do a round of checks to see who's published for that milestone, but to be honest, I'd say anyone who's on RC2 already is way ahead of the curve and won't have trouble upgrading, so we might as well skip tracking that specifically and focus on the remaining libraries.

**Comment by kubukoz** [comment]
Update, 3.0.0 is out! After I'm done with the migration guide I'll start updating the table to track projects moving to that.

**Comment by miriarte** [comment]
fs2 (https://github.com/typelevel/fs2/releases/tag/v3.0.0) and http4s (https://github.com/http4s/http4s/releases/tag/v1.0.0-M20) already released versions using CE 3.0.0

**Comment by kubukoz** [comment]
12 libraries from the list are released ;)

**Comment by rkrzewski** [comment]
[elastic4s](https://github.com/sksamuel/elastic4s/) has been upgraded https://github.com/sksamuel/elastic4s/pull/2351, pending relelease

**Comment by kubukoz** [comment]
22 libraries from the list released.

**Comment by kubukoz** [comment]
https://github.com/cb372/cats-retry/pull/305 gonna list this next

**Comment by kbreidenbachRN** [comment]
We're using `cats-effect` in production along with `scanamo-cats-effect` for DynamoDB. At the moment I have no information on when `scanamo-cats-effect` will be updated to use CE3 so this is holding some projects back

**Comment by kubukoz** [comment]
Scanamo seems to be using Monix in the ~core~ cats module, so I'd say it'll be blocked on that until Monix gets there :/ 

it's in `provided` scope, but there's a `toIterant` function that touches both the CE and Monix worlds, so it seems pretty dangerous to upgrade now.

**Comment by kubukoz** [comment]
@kbreidenbachRN as a temporary alternative, I would suggest creating your own `scanamo-cats` library using the current `scanamo-cats-effect` code as the base (it's apache 2 licensed so you're allowed to do that). The migration itself should be straightforward, mostly changing the `async` call to `Async[F].fromCompletableFuture`.

**Comment by kubukoz** [comment]
Actually, hold that thought... it's just one, one-line method for the Monix interop - maybe this can just be dropped? The fs2 integration is also very thin and probably could be removed as well to avoid making the dependency graph unnecessarily deep. @regiskuckaertz I would appreciate your input here! 🙏

**Comment by kbreidenbachRN** [comment]
@kubukoz As we are using FS2 as well, I just took your first suggestion: forked, made the `Async[F].fromCompletableFuture` change and put a SNAPSHOT version in our artifactory repo. We should be in prod with a few CE3 updated apps in a couple of days. Thanks for the help!

**Comment by vasyl-zhurba** [comment]
[circe-derivation](https://github.com/circe/circe-derivation)

**Comment by kubukoz** [comment]
> [circe-derivation](https://github.com/circe/circe-derivation)

This project doesn't use cats-effect.

**Comment by vasyl-zhurba** [comment]
Sorry, my mistake.
I was trying to upgrade my project to cats-effect 3 and got the following errors
```
* org.http4s:http4s-core_2.13:1.0.0-M23 (early-semver) is selected over {0.21.23, 0.21.23}
[error] 	    +- org.http4s:http4s-jawn_2.13:1.0.0-M23              (depends on 1.0.0-M23)
[error] 	    +- org.http4s:http4s-blaze-core_2.13:1.0.0-M23        (depends on 1.0.0-M23)
[error] 	    +- org.http4s:http4s-circe_2.13:1.0.0-M23             (depends on 1.0.0-M23)
[error] 	    +- org.http4s:http4s-client_2.13:1.0.0-M23            (depends on 1.0.0-M23)
[error] 	    +- org.http4s:http4s-server_2.13:0.21.23              (depends on 0.21.23)
[error] 	    +- org.http4s:http4s-dsl_2.13:0.21.23                 (depends on 0.21.23)
[error]
[error] 	* org.http4s:http4s-client_2.13:1.0.0-M23 (early-semver) is selected over 0.23.0-M1
[error] 	    +- org.http4s:http4s-blaze-client_2.13:1.0.0-M23      (depends on 1.0.0-M23)
[error] 	    +- com.softwaremill.sttp.client3:http4s-backend_2.13:3.3.5 (depends on 0.23.0-M1)
[error]
[error] 	* org.http4s:http4s-blaze-core_2.13:1.0.0-M23 (early-semver) is selected over 0.21.23
[error] 	    +- org.http4s:http4s-blaze-client_2.13:1.0.0-M23      (depends on 1.0.0-M23)
[error] 	    +- org.http4s:http4s-blaze-server_2.13:0.21.23        (depends on 0.21.23)
[error]
[error] 	* co.fs2:fs2-io_2.13:3.0.4 (early-semver) is selected over {2.5.6, 3.0.3, 3.0.3}
[error] 	    +- org.http4s:http4s-core_2.13:1.0.0-M23              (depends on 3.0.4)
[error] 	    +- org.tpolecat:doobie-postgres_2.13:1.0.0-M5         (depends on 3.0.3)
[error] 	    +- com.softwaremill.sttp.shared:fs2_2.13:1.2.5        (depends on 3.0.3)
[error] 	    +- com.softwaremill.sttp.shared:fs2-ce2_2.13:1.2.5    (depends on 2.5.6)
```
What did I miss? 
Looks like sttp-http4s-backend v.3.3.5 still depends on http4s v.0.23
I couldn't find explicit usage of htt4s v0.21.23 in my project.

**Comment by Daenyth** [comment]
Fetch is now updated: https://github.com/47degrees/fetch/releases/tag/v2.0.0

**Comment by andreamarcolin** [comment]
Catbird has been published for CE3 (catbird-effect3 module, version 21.5.0 https://github.com/typelevel/catbird/pull/267), also unblocking Tapir (https://github.com/softwaremill/tapir/pull/1154)

**Comment by ghostdogpr** [comment]
Caliban interop for cats-effect 3 and http4s 0.23 was just released ✅

**Comment by henricook** [comment]
@kubukoz When you get a moment would you mind updating the table? Thanks for your work thus far

**Comment by henricook** [comment]
Hey @kubukoz - would you mind updating this table when you get a second? Thanks for your work thus far

**Comment by Daenyth** [comment]
circe-fs2 is released https://github.com/circe/circe-fs2/releases/tag/v0.14.0

**Comment by kubukoz** [comment]
@djspiewak think we can close this?

**Comment by djspiewak** [comment]
I believe indeed we can. Thank you for all your hard work on this, particularly two years ago!
